internal/epoch/epoch_manager.go

```diff
@@ -1067,6 +1067,57 @@ func getCompactedEpochRange(cs CurrentSnapshot) (intRange, error) {
 	return getKeyRange(cs.SingleEpochCompactionSets)
 }
 
+// MaybeCompactSingleEpoch compacts the oldest epoch that is elegible for
+// compaction if there is one.
+func (e *Manager) MaybeCompactSingleEpoch(ctx context.Context) error {
+	cs, err := e.committedState(ctx, 0)
+	if err != nil {
+		return err
+	}
+
+	uncompacted, err := getUncompactedEpochRange(cs)
+	if err != nil {
+		return err
+	}
+
+	if uncompacted.isEmpty() {
+		e.log.Debug("empty uncompacted epoch range, not compacting")
+		return nil
+	}
+
+	compacted, err := getCompactedEpochRange(cs)
+	if err != nil {
+		return err
+	}
+
+	if uncompacted.lo < compacted.hi {
+		uncompacted.lo = compacted.hi // avoid compacting already compacted epoch
```

Comment:
What scenario would lead to this (if any)? Committed state in an inconsistent/transitory state? Racing with other compactions?

Reply:
Good question. Let me dig again.

IIRC:
- It is what you mention. this function is not assuming it will be called sequentially, in isolation, etc, and that instead there may be cross-client concurrency. This is even if the intent is to only call this from maintenance to guarantee serial execution.
- It is also due to the way the "committed state" is loaded. Again, need to double check.

Reply:
The implementation changed to account for the case where the list of blobs for some uncompacted epochs is not included in the _"current snapshot"_ (`cs`).

See https://github.com/kopia/kopia/pull/3728/files#diff-0fc9e0e50903701c369ab11147ea2f21b5c9756abac3299c28aeab8bc628ce41